# Bank Customer Churn Prediction: An End-to-End Machine Learning Experimentation Pipeline

## Project Overview

Customer churn prediction is one of the most important real-world applications of Machine Learning in the banking and financial sector. Retaining existing customers is significantly more cost-effective than acquiring new ones, making churn analysis a critical business problem for modern organizations.

This project focuses on building a complete end-to-end Machine Learning experimentation pipeline for predicting customer churn using banking customer data. Rather than training a single model and reporting accuracy, this project emphasizes analytical thinking, preprocessing decisions, model evaluation, experimentation, and understanding the impact of different Machine Learning techniques on predictive performance.

---

## Dataset

The dataset used in this project is the Bank Customer Churn Prediction dataset, which contains customer demographic information, banking activity, and churn status used for classification analysis.

Dataset Source:  
[Kaggle - Bank Customer Churn Prediction Dataset](https://www.kaggle.com/datasets/shantanudhakadd/bank-customer-churn-prediction)

---

## The project includes:

- Data preprocessing and feature engineering
- Outlier analysis and business-driven preprocessing decisions
- Categorical feature encoding
- Splitting the dataset into the Train set and Test set
- Feature scaling experiments
- Class imbalance handling techniques
- Model training and evaluation
- Cross-validation for performance stability analysis
- Comparative experimentation across multiple Machine Learning models

A major focus of this project is understanding how preprocessing choices affect model behavior and performance in imbalanced classification problems. Different strategies such as Standard Scaling, Robust Scaling, class weighting, and imbalance handling techniques were explored and critically evaluated rather than applied blindly.

The experiments demonstrate the importance of looking beyond accuracy when evaluating classification models. Metrics such as **Precision**, **Recall**, **F1-Score**, **ROC-AUC**, and **Confusion Matrix** analysis were used to assess model effectiveness, particularly for detecting churned customers.

This project was designed not only to build predictive models, but also to develop a deeper understanding of practical Machine Learning workflows used in real-world industry environments.

---

## Project Objectives

- Predict whether a customer is likely to churn
- Analyze the impact of preprocessing techniques on model performance
- Evaluate multiple classification algorithms
- Investigate class imbalance handling methods
- Compare model performance using multiple evaluation metrics
- Build a reproducible and professional Machine Learning experimentation workflow

---

## Technologies and Libraries

- Python
- NumPy
- Pandas
- Scikit-learn
- Imbalanced-learn

---

## Machine Learning Concepts Explored

- Classification
- Logistic Regression
- K-Nearest Neighbors (KNN)
- Support Vector Machine SVM
- Naive Bayes
- Decision Trees
- Random Forest
- One-Hot Encoding
- Feature Scaling
- StandardScaler vs RobustScaler
- Outlier Analysis
- Class Imbalance Handling
- Cross Validation
- Model Evaluation Metrics
- ROC-AUC Analysis



# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## Data Cleaning Check Template
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv("Churn_Modelling.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB
None

SHAPE OF DATA
Rows: 10000, Columns: 14

MISSING VALUES
Se

# **Outlier Analysis**
Outlier detection was performed using the IQR method on the numerical features.

### Features with Detected Outliers

- `CreditScore`
- `Age`
- `NumOfProducts`
- `Exited`


### CreditScore: 15 outliers

In [4]:
# Calculate fences for year
Q1 = dataset['CreditScore'].quantile(0.25)
Q3 = dataset['CreditScore'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
year_outliers = dataset[(dataset['CreditScore'] < lower_fence) | (dataset['CreditScore'] > upper_fence)]

# Display the rows
print(year_outliers)

      RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
7             8    15656148     Obinna          376   Germany  Female   29   
942         943    15804586        Lin          376    France  Female   46   
1193       1194    15779947     Thomas          363     Spain  Female   28   
1405       1406    15612494  Panicucci          359    France  Female   44   
1631       1632    15685372   Azubuike          350     Spain    Male   54   
1838       1839    15758813   Campbell          350   Germany    Male   39   
1962       1963    15692416  Aikenhead          358     Spain  Female   52   
2473       2474    15679249       Chou          351   Germany  Female   57   
2579       2580    15597896   Ozoemena          365   Germany    Male   30   
8154       8155    15791533     Ch'ien          367     Spain    Male   42   
8723       8724    15803202  Onyekachi          350    France    Male   51   
8762       8763    15765173        Lin          350    France  F

In [5]:
print(dataset["CreditScore"])

0       619
1       608
2       502
3       699
4       850
       ... 
9995    771
9996    516
9997    709
9998    772
9999    792
Name: CreditScore, Length: 10000, dtype: int64


### Age: 359 outliers

In [6]:
# Calculate fences for year
Q1 = dataset['Age'].quantile(0.25)
Q3 = dataset['Age'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
year_outliers = dataset[(dataset['Age'] < lower_fence) | (dataset['Age'] > upper_fence)]

# Display the rows
print(year_outliers)

      RowNumber  CustomerId      Surname  CreditScore Geography  Gender  Age  \
58           59    15623944        T'ien          511     Spain  Female   66   
85           86    15805254      Ndukaku          652     Spain  Female   75   
104         105    15804919     Dunbabin          670     Spain  Female   65   
158         159    15589975      Maclean          646    France  Female   73   
181         182    15789669         Hsia          510    France    Male   65   
...         ...         ...          ...          ...       ...     ...  ...   
9753       9754    15705174    Chiedozie          656   Germany    Male   68   
9765       9766    15777067       Thomas          445    France    Male   64   
9832       9833    15814690  Chukwujekwu          595   Germany  Female   64   
9894       9895    15704795        Vagin          521    France  Female   77   
9936       9937    15653037        Parks          609    France    Male   77   

      Tenure    Balance  NumOfProducts 

In [7]:
print(dataset["Age"])

0       42
1       41
2       42
3       39
4       43
        ..
9995    39
9996    35
9997    36
9998    42
9999    28
Name: Age, Length: 10000, dtype: int64


### NumOfProducts: 60 outliers

In [8]:
# Calculate fences for year
Q1 = dataset['NumOfProducts'].quantile(0.25)
Q3 = dataset['NumOfProducts'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
year_outliers = dataset[(dataset['NumOfProducts'] < lower_fence) | (dataset['NumOfProducts'] > upper_fence)]

# Display the rows
print(year_outliers)

      RowNumber  CustomerId           Surname  CreditScore Geography  Gender  \
7             8    15656148            Obinna          376   Germany  Female   
70           71    15703793        Konovalova          738   Germany    Male   
1254       1255    15610383        Dumetolisa          628    France  Female   
1469       1470    15670374            Wright          819   Germany  Female   
1488       1489    15625824         Kornilova          596     Spain    Male   
1701       1702    15605279           Francis          792    France    Male   
1876       1877    15664754            Steele          640   Germany    Male   
2124       2125    15603851            Galkin          704    France    Male   
2196       2197    15791700  Ugochukwutubelum          773   Germany    Male   
2285       2286    15783615             Ramos          630   Germany    Male   
2462       2463    15704442           Fleming          672    France  Female   
2499       2500    15634974           Se

In [9]:
print(dataset["NumOfProducts"])

0       1
1       1
2       3
3       2
4       1
       ..
9995    2
9996    1
9997    1
9998    2
9999    1
Name: NumOfProducts, Length: 10000, dtype: int64


### Exited: 2037 outliers

In [10]:
# Calculate fences for year
Q1 = dataset['Exited'].quantile(0.25)
Q3 = dataset['Exited'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
year_outliers = dataset[(dataset['Exited'] < lower_fence) | (dataset['Exited'] > upper_fence)]

# Display the rows
print(year_outliers)

      RowNumber  CustomerId      Surname  CreditScore Geography  Gender  Age  \
0             1    15634602     Hargrave          619    France  Female   42   
2             3    15619304         Onio          502    France  Female   42   
5             6    15574012          Chu          645     Spain    Male   44   
7             8    15656148       Obinna          376   Germany  Female   29   
16           17    15737452        Romeo          653   Germany    Male   58   
...         ...         ...          ...          ...       ...     ...  ...   
9981       9982    15672754     Burbidge          498   Germany    Male   42   
9982       9983    15768163      Griffin          655   Germany  Female   46   
9991       9992    15769959  Ajuluchukwu          597    France  Female   53   
9997       9998    15584532          Liu          709    France  Female   36   
9998       9999    15682355    Sabbatini          772   Germany    Male   42   

      Tenure    Balance  NumOfProducts 

In [11]:
print(dataset["Exited"])

0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64


### Decision About Outliers

The detected outliers were intentionally retained instead of being removed because these values represent realistic customer behavior rather than data entry errors.

#### CreditScore
Extreme credit scores are valid in banking datasets and may contain important information about customer financial behavior and churn patterns. Removing these observations could reduce the model’s ability to learn from high-risk or low-risk customers.

#### Age
Older customers were detected as statistical outliers due to the IQR method. However, these customers are valid observations and may have significantly different churn behavior compared to younger customers. Therefore, they were preserved.

#### NumOfProducts
Customers with a high number of banking products were identified as outliers. These cases are meaningful from a business perspective because product usage can strongly influence customer retention and churn.

#### Exited
The `Exited` feature represents the target variable of the classification problem, where:

- `0` = Customer stayed
- `1` = Customer churned

Although the IQR method detected many values as outliers, this result is not meaningful because `Exited` is a binary categorical variable rather than a continuous numerical feature.

Outlier detection methods such as IQR are designed for continuous numerical distributions and should not be applied to classification labels.

Therefore, no outlier handling was performed on the `Exited` column, and all observations were retained.

### Additional Notes
- Identifier columns such as:
  - `RowNumber`
  - `CustomerId`

  were removed from the dataset because they do not provide predictive value.

- `Surname`

The `Surname` column was removed from the dataset because it represents a high cardinality categorical feature with a large number of unique customer names.

Customer surnames are unlikely to have a direct causal relationship with churn behavior and may introduce noise into the model.

Additionally, encoding this feature could create unnecessary dimensionality or artificial relationships between categories without providing meaningful predictive value.

Therefore, the `Surname` feature was excluded from the modeling process.

### Conclusion

The outliers were retained to preserve potentially valuable business information and avoid removing legitimate customer behavior patterns. Tree-based models such as Random Forest are generally robust to outliers, making aggressive outlier removal unnecessary for this project.

# **Select Independent Variable (X) and Dependent Variable (y)**

Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [12]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset.iloc[:, 3:-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset.iloc[:, -1].values

In [13]:
print(X[0:10, :])

[[619 'France' 'Female' 42 2 0.0 1 1 1 101348.88]
 [608 'Spain' 'Female' 41 1 83807.86 1 0 1 112542.58]
 [502 'France' 'Female' 42 8 159660.8 3 1 0 113931.57]
 [699 'France' 'Female' 39 1 0.0 2 0 0 93826.63]
 [850 'Spain' 'Female' 43 2 125510.82 1 1 1 79084.1]
 [645 'Spain' 'Male' 44 8 113755.78 2 1 0 149756.71]
 [822 'France' 'Male' 50 7 0.0 2 1 1 10062.8]
 [376 'Germany' 'Female' 29 4 115046.74 4 1 0 119346.88]
 [501 'France' 'Male' 44 4 142051.07 2 0 1 74940.5]
 [684 'France' 'Male' 27 2 134603.88 1 1 1 71725.73]]


# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [14]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False, drop='first'),[1,2])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

| Model               | Drop dummy column? | Why                       |
| ------------------- | ------------------ | ------------------------- |
| Logistic Regression | Yes                | multicollinearity         |
| SVM                 | Recommended        | stability                 |
| KNN                 | Optional           | distance redundancy       |
| Naive Bayes         | No                 | independence assumption   |
| Decision Tree       | No                 | splits ignore correlation |
| Random Forest       | No                 | robust ensemble           |


In [15]:
print(X[0:10, :])
#france  0.0 0.0
#spain   0.0 1.0
#germany 1.0 0.0
#female  0.0
#male    1.0

[[0.0 0.0 0.0 619 42 2 0.0 1 1 1 101348.88]
 [0.0 1.0 0.0 608 41 1 83807.86 1 0 1 112542.58]
 [0.0 0.0 0.0 502 42 8 159660.8 3 1 0 113931.57]
 [0.0 0.0 0.0 699 39 1 0.0 2 0 0 93826.63]
 [0.0 1.0 0.0 850 43 2 125510.82 1 1 1 79084.1]
 [0.0 1.0 1.0 645 44 8 113755.78 2 1 0 149756.71]
 [0.0 0.0 1.0 822 50 7 0.0 2 1 1 10062.8]
 [1.0 0.0 0.0 376 29 4 115046.74 4 1 0 119346.88]
 [0.0 0.0 1.0 501 44 4 142051.07 2 0 1 74940.5]
 [0.0 0.0 1.0 684 27 2 134603.88 1 1 1 71725.73]]


### Categorical Feature Encoding

Categorical variables were transformed into numerical representations using One-Hot Encoding.

#### Geography
The `Geography` feature was encoded using One-Hot Encoding. Since the feature contains three categories (`France`, `Spain`, and `Germany`), two binary columns were created while one category was automatically dropped to avoid the dummy variable trap and multicollinearity.

Encoding representation:

| Germany | Spain | Geography |
|---|---|---|
| 0 | 0 | France |
| 0 | 1 | Spain |
| 1 | 0 | Germany |

#### Gender
The `Gender` feature was also transformed using One-Hot Encoding. Since it contains only two categories (`Female` and `Male`), a single binary column was generated after dropping one category.

Encoding representation:

| Gender |
|---|
| 0 → Female |
| 1 → Male |

These transformations convert categorical variables into numerical representations suitable for machine learning algorithms.

# **Splitting the dataset into the Train set and Test set**

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [17]:
print(X_train[0:10, :])

[[1.0 0.0 0.0 579 39 5 117833.3 3 0 0 5831.0]
 [0.0 0.0 0.0 750 32 5 0.0 2 1 0 95611.47]
 [0.0 1.0 0.0 729 34 9 53299.96 2 1 1 42855.97]
 [0.0 1.0 1.0 689 38 5 75075.14 1 1 1 8651.92]
 [0.0 0.0 1.0 605 52 7 0.0 2 1 1 173952.5]
 [0.0 0.0 0.0 667 37 9 71786.9 2 1 1 67734.79]
 [0.0 0.0 1.0 673 65 0 0.0 1 1 1 85733.33]
 [0.0 0.0 1.0 724 31 5 0.0 1 1 0 134889.95]
 [0.0 0.0 0.0 731 38 10 123711.73 2 1 0 171340.68]
 [0.0 1.0 0.0 484 39 5 0.0 2 1 1 175224.12]]


In [18]:
print(X_test[0:10, :])

[[1.0 0.0 0.0 597 35 8 131101.04 1 1 1 192852.67]
 [0.0 0.0 0.0 523 40 2 102967.41 1 1 0 128702.1]
 [0.0 1.0 0.0 706 42 8 95386.82 1 1 1 75732.25]
 [0.0 0.0 1.0 788 32 4 112079.58 1 0 0 89368.59]
 [1.0 0.0 1.0 706 38 5 163034.82 2 1 1 135662.17]
 [0.0 1.0 0.0 670 57 3 175575.95 2 1 0 99061.75]
 [0.0 1.0 1.0 590 34 0 65812.35 2 0 1 160346.3]
 [0.0 1.0 0.0 636 29 6 157576.47 2 1 1 101102.39]
 [0.0 0.0 0.0 598 64 9 0.0 1 0 1 13181.37]
 [0.0 0.0 0.0 456 63 1 165350.61 2 0 0 140758.07]]


In [19]:
print(y_train)

[1 0 0 ... 0 0 1]


In [20]:
print(y_test)

[0 1 0 ... 0 0 0]


# **Feature Scaling Using StandardScaler**

Note: I explained Feature Scaling earlier.

If you want to understand when and why to use feature scaling, you can check it here: [Go to  When is Feature Scaling Necessary in Regression Models? Section](https://github.com/Hazem1695/ml-concept-briefs)

In [21]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train[:, [3,4,5,6,7,10]] = sc.fit_transform(X_train[:, [3,4,5,6,7,10]])
X_test[:, [3,4,5,6,7,10]] = sc.transform(X_test[:, [3,4,5,6,7,10]])

In [22]:
print(X_train[0:10, :])

[[1.0 0.0 0.0 -0.7355070560983401 0.0152657115778632 0.00886037125109047
  0.6731600273055683 2.5350339381815266 0 0 -1.640809940586845]
 [0.0 0.0 0.0 1.0244271940236245 -0.652609169953666 0.00886037125109047
  -1.2077241651131092 0.8042415361458916 1 0 -0.07927152135221784]
 [0.0 1.0 0.0 0.808294917692857 -0.46178777523037196 1.3932933792339848
  -0.3569370597655732 0.8042415361458916 1 1 -0.9968401203574622]
 [0.0 1.0 1.0 0.39661439134853776 -0.08014498578378383
  0.00886037125109047 -0.009356266822735035 -0.9265508658897434 1 1
  -1.5917460967732924]
 [0.0 0.0 1.0 -0.4679147139745326 1.2556047772792747 0.7010768752425377
  -1.2077241651131092 0.8042415361458916 1 1 1.2833024120144987]
 [0.0 0.0 0.0 0.17019010185916217 -0.17555568314543085 1.3932933792339848
  -0.0618439645862603 0.8042415361458916 1 1 -0.5641264883702068]
 [0.0 0.0 1.0 0.23194218081081006 2.495943842980686 -1.7216808887275274
  -1.2077241651131092 -0.9265508658897434 1 1 -0.25108054753105535]
 [0.0 0.0 1.0 0.7568348

In [23]:
print(X_test[0:10, :])

[[1.0 0.0 0.0 -0.5502508192433965 -0.3663770778687249 1.0471851272382613
  0.8849429650245407 -0.9265508658897434 1 1 1.6120302699224127]
 [0.0 0.0 0.0 -1.311859792980387 0.11067640893951022 -1.0294643847360803
  0.4358670251810849 -0.9265508658897434 1 0 0.4962689058885611]
 [0.0 1.0 0.0 0.5715786150448734 0.30149780366280426 1.0471851272382613
  0.31486377821942724 -0.9265508658897434 1 1 -0.425027850923865]
 [0.0 0.0 1.0 1.4155236940507279 -0.652609169953666 -0.33724788074463313
  0.5813177297920199 -0.9265508658897434 0 0 -0.18785300869207555]
 [1.0 0.0 1.0 0.5715786150448734 -0.08014498578378383 0.00886037125109047
  1.394677836289079 0.8042415361458916 1 1 0.6173243726906659]
 [0.0 1.0 0.0 0.20106614133498613 1.7326582640875097 -0.6833561327403567
  1.5948624460052427 0.8042415361458916 1 0 -0.01926131225978551]
 [0.0 1.0 1.0 -0.6222949113536523 -0.46178777523037196
  -1.7216808887275274 -0.15721120501672425 0.8042415361458916 0 1
  1.0466517903393149]
 [0.0 1.0 0.0 -0.1488623060

# **Building Logistic Regression Classification Model**

## Training the Building Logistic Regression Classification model on the Training set

In [24]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(class_weight = 'balanced', random_state = 0)
classifier.fit(X_train,y_train)

LogisticRegression(class_weight='balanced', random_state=0)

### Handling Class Imbalance

The dataset contains an imbalanced target distribution, where the number of non-churned customers is significantly higher than churned customers.

To reduce bias toward the majority class, the `class_weight='balanced'` parameter was applied during model training. This automatically adjusts class weights inversely proportional to class frequencies, giving more importance to minority class samples (`Exited = 1`).

This approach helps improve the model’s ability to detect churned customers and increases recall for the minority class.

---

### Difference Between Handling Class Imbalance Methods

| Method        | Action                             |
| ------------- | ---------------------------------- |
| Class Weights | changes learning importance        |
| SMOTE         | creates synthetic minority samples |
| Undersampling | removes majority samples           |


# **Preticting the Test set results**

In [25]:
y_pred = classifier.predict(X_test)

print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1)[0:100,:])

[[1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [1 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 0]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [1 0]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]]


# **Evaluating the Model Performance**

In [29]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score
from sklearn.model_selection import cross_val_score

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nROC-AUC Score:")
print(roc_auc_score(y_test, y_pred))


accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=10)
print("\nMean Accuracy:")
print(accuracies.mean())
print("\nStandard Deviation:")
print(accuracies.std())



Confusion Matrix:
[[1385  606]
 [ 136  373]]

Accuracy Score:
0.7032

Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.70      0.79      1991
           1       0.38      0.73      0.50       509

    accuracy                           0.70      2500
   macro avg       0.65      0.71      0.65      2500
weighted avg       0.80      0.70      0.73      2500


ROC-AUC Score:
0.7142198833848585

Mean Accuracy:
0.7081333333333334

Standard Deviation:
0.010404272626613002


| ROC-AUC | Quality         |
| ------- | --------------- |
| 0.5     | random          |
| 0.6     | weak            |
| 0.7     | acceptable/good |
| 0.8     | strong          |


# Model Comparison Analysis

This experiment evaluates the impact of handling class imbalance using `class_weight='balanced'` and compares the effect of two different feature scaling techniques: `StandardScaler` and `RobustScaler`.

The main objective was not only to maximize overall accuracy, but also to improve the model’s ability to correctly identify the minority class, which is the most important challenge in imbalanced classification problems such as customer churn prediction.

---

# 1. Baseline Model

## Without `class_weight='balanced'` using `StandardScaler`

The baseline model achieved the highest overall accuracy score at **80.68%**, which may initially appear to indicate strong performance. However, deeper evaluation reveals that the model struggled significantly with detecting the minority class.

The confusion matrix shows that the model correctly identified most majority-class samples, achieving a very high recall of **95%** for class `0`. However, it only detected **116** out of **509** minority-class samples, resulting in a minority-class recall of just **23%**.

This behavior demonstrates a common issue in imbalanced datasets: the model becomes biased toward the dominant class and achieves high accuracy simply by predicting the majority class more frequently.

Although the weighted metrics appear acceptable, the minority-class performance remained weak:

* Minority-class precision: **56%**
* Minority-class recall: **23%**
* Minority-class F1-score: **32%**
* Receiver Operating Characteristic - Area Under Curve score: **0.591**

The low Receiver Operating Characteristic - Area Under Curve score indicates limited class separation capability and poor balance between true positive and false positive predictions.

---

# 2. Balanced Model

## With `class_weight='balanced'` using `StandardScaler`

After introducing `class_weight='balanced'`, the model became significantly more effective at identifying minority-class samples.

The minority-class recall increased dramatically from **23%** to **73%**, meaning the model successfully captured a much larger portion of positive cases. This improvement is highly important in classification tasks where missing minority-class samples carries a high business cost.

The confusion matrix confirms this improvement:

* False negatives decreased from **393** to **136**
* True positives increased from **116** to **373**

Although overall accuracy dropped from **80.68%** to **70.32%**, this trade-off is expected in imbalanced classification tasks because the model is no longer overly biased toward the majority class.

Additional improvements include:

* Receiver Operating Characteristic - Area Under Curve score improved from **0.591** to **0.714**
* Macro-average recall improved substantially
* Minority-class F1-score increased from **32%** to **50%**

This model demonstrates a much more balanced learning behavior and provides stronger generalization for real-world imbalanced classification scenarios.

---

# 3. Balanced Model

## With `class_weight='balanced'` using `RobustScaler`

The model trained with `RobustScaler` produced nearly identical results to the model trained with `StandardScaler`.

Performance differences were extremely small:

* Accuracy improved slightly from **70.32%** to **70.36%**
* Receiver Operating Characteristic - Area Under Curve score improved marginally from **0.7142** to **0.7145**
* Cross-validation mean accuracy improved slightly
* Standard deviation increased marginally

The confusion matrix and classification report remained effectively unchanged.

This indicates that the dataset was not heavily affected by extreme outliers, since `RobustScaler` is specifically designed to reduce the influence of outlier values. As a result, replacing `StandardScaler` with `RobustScaler` did not lead to meaningful performance improvements.

---

# Performance Comparison Table

| Configuration                                  | Accuracy | Minority Recall | Minority Precision | Minority F1-Score | Receiver Operating Characteristic - Area Under Curve | Mean Cross Validation Accuracy |
| ---------------------------------------------- | -------- | --------------- | ------------------ | ----------------- | ---------------------------------------------------- | ------------------------------ |
| Without Balanced Class Weight + StandardScaler | 80.68%   | 23%             | 56%                | 32%               | 0.591                                                | 80.85%                         |
| Balanced Class Weight + StandardScaler         | 70.32%   | 73%             | 38%                | 50%               | 0.714                                                | 70.81%                         |
| Balanced Class Weight + RobustScaler           | 70.36%   | 73%             | 38%                | 50%               | 0.714                                                | 70.83%                         |

---

# Confusion Matrix Comparison

| Configuration                                  | True Negatives | False Positives | False Negatives | True Positives |
| ---------------------------------------------- | -------------- | --------------- | --------------- | -------------- |
| Without Balanced Class Weight + StandardScaler | 1901           | 90              | 393             | 116            |
| Balanced Class Weight + StandardScaler         | 1385           | 606             | 136             | 373            |
| Balanced Class Weight + RobustScaler           | 1386           | 605             | 136             | 373            |

---

# Cross Validation Stability Comparison

| Configuration                                  | Mean Accuracy | Standard Deviation |
| ---------------------------------------------- | ------------- | ------------------ |
| Without Balanced Class Weight + StandardScaler | 0.8085        | 0.0089             |
| Balanced Class Weight + StandardScaler         | 0.7081        | 0.0104             |
| Balanced Class Weight + RobustScaler           | 0.7083        | 0.0106             |

---

# Insights

## 1. Accuracy Alone Was Misleading

The baseline model achieved the highest accuracy, but failed to identify most minority-class samples. This demonstrates why accuracy should not be the primary evaluation metric for imbalanced classification problems.

---

## 2. Balanced Class Weight Significantly Improved Minority Detection

Applying `class_weight='balanced'` transformed the model behavior:

* Minority recall increased from **23%** to **73%**
* False negatives were reduced substantially
* Receiver Operating Characteristic - Area Under Curve score improved considerably

This indicates that class balancing helped the model learn decision boundaries that better represent both classes instead of favoring the majority class.

---

## 3. The Model Became More Balanced Rather Than More Accurate

After balancing, the model sacrificed some majority-class accuracy in exchange for dramatically better minority detection.

This trade-off is often desirable in real-world applications such as:

* Customer churn prediction
* Fraud detection
* Medical diagnosis
* Risk prediction systems

where missing positive cases can be more costly than generating additional false positives.

---

## 4. RobustScaler Did Not Provide Significant Improvement

The nearly identical performance between `StandardScaler` and `RobustScaler` suggests that:

* Outliers were not a major issue in the dataset
* The model performance was influenced more by class imbalance than scaling strategy

This experiment highlights the importance of understanding the actual bottleneck affecting model performance instead of applying preprocessing techniques blindly.

---

# Conclusion

The final selected configuration was:

## `class_weight='balanced'` with `StandardScaler`

Although this configuration achieved lower overall accuracy compared to the baseline model, it provided substantially better minority-class detection and a much stronger Receiver Operating Characteristic - Area Under Curve score.

The experiment demonstrates an important machine learning principle:

> A model with lower accuracy can still be significantly better if it improves the ability to correctly classify the minority class.

From a practical machine learning perspective, the balanced model is more reliable and more suitable for real-world imbalanced classification tasks because it prioritizes meaningful class detection rather than inflated accuracy scores driven by majority-class dominance.

This experiment also demonstrates a strong understanding of:

* Imbalanced classification challenges
* Trade-offs between evaluation metrics
* Proper interpretation of confusion matrices
* Model evaluation beyond accuracy
* The effect of preprocessing techniques on classification performance
* Practical decision-making in machine learning experimentation
